src/aleph/http/multipart.clj

```diff
@@ -146,3 +158,94 @@
           (.addBodyHttpData encoder attr))))
     (let [req' (.finalizeRequest encoder)]
       [req' (when (.isChunked encoder) encoder)])))
+
+(defmulti http-data->map
+  (fn [^InterfaceHttpData data]
+    (.getHttpDataType data)))
+
+(defmethod http-data->map InterfaceHttpData$HttpDataType/Attribute
+  [^Attribute attr]
+  (let [content (.getValue attr)]
+    {:part-name (.getName attr)
+     :content content
+     :name nil
+     :charset (-> attr .getCharset .toString)
+     :mime-type nil
+     :transfer-encoding nil
+     :memory? (.isInMemory attr)
+     :file? false
+     :file nil
+     :size (count content)}))
+
+(defmethod http-data->map InterfaceHttpData$HttpDataType/FileUpload
+  [^FileUpload data]
+  (let [memory? (.isInMemory data)]
+    {:part-name (.getName data)
+     :content (when memory?
+                (bs/to-input-stream (netty/acquire (.content data))))
+     :name (.getFilename data)
+     :charset (-> data .getCharset .toString)
+     :mime-type (.getContentType data)
+     :transfer-encoding (.getContentTransferEncoding data)
+     :memory? memory?
+     :file? true
+     :file (when-not memory? (.getFile data))
+     :size (.length data)}))
+
+(defn- read-attributes [^HttpPostRequestDecoder decoder parts]
+  (while (.hasNext decoder)
+    (s/put! parts (http-data->map (.next decoder)))))
+
+(defn decode-request
+  "Takes a ring request and returns a manifold stream which yields
+   parts of the mutlipart/form-data encoded body. In case the size of
+   a part content exceeds `:memory-limit` limit (16KB by default),
+   corresponding payload would be written to a temp file. Check `:memory?`
+   flag to know whether content might be read directly from `:content` or
+   should be fetched from the file specified in `:file`.
+
+   Note, that if your handler works with multipart requests only,
+   it's better to set `:raw-stream?` to `true` to avoid additional
+   input stream coercion."
+  ([req] (decode-request req {}))
+  ([{:keys [body] :as req}
+    {:keys [body-buffer-size
+            memory-limit]
+     :or {body-buffer-size 65536
+          memory-limit DefaultHttpDataFactory/MINSIZE}}]
+   (let [body (if (s/stream? body)
+                body
+                (netty/to-byte-buf-stream body body-buffer-size))
+         destroyed? (atom false)
+         req' (http-core/ring-request->netty-request req)
+         factory (DefaultHttpDataFactory. (long memory-limit))
+         decoder (HttpPostRequestDecoder. factory req')
+         parts (s/stream)]
+
+     ;; on each HttpContent chunk, put it into the decoder
+     ;; and resume our attempts to get the next attribute available
+     (s/connect-via
+      body
+      (fn [chunk]
+        (let [content (DefaultHttpContent. chunk)]
+          (.offer decoder content)
+          (read-attributes decoder parts)
+          ;; TODO(kachayev): I'm still not sure this is
```

Comment:
If we're feeling paranoid about this, we could just coerce it to a ByteBuffer and then back to a ByteBuf (now on-heap).  I don't think the memory copy overhead is worth worrying about.

Reply:
Nope, at all. I'm going just remove this comment as it was done initially. I've checked the implementation on the netty side and this sequence of operations seems pretty reasonable. More than that, I do not expect a lot of changes in those decoders in any foreseeable future.